sdk/core/azure-core/azure/core/pipeline/policies/_retry.py

```diff
@@ -296,12 +296,18 @@ def _configure_timeout(self, request, absolute_timeout, is_response_error):
             if is_response_error:
                 raise ServiceResponseTimeoutError('Response timeout')
             raise ServiceRequestTimeoutError('Request timeout')
+
+        # if connection_timeout is already set, ensure it doesn't exceed absolute_timeout
         connection_timeout = request.context.options.get('connection_timeout')
         if connection_timeout:
-            req_timeout = min(connection_timeout, absolute_timeout)
-        else:
-            req_timeout = absolute_timeout
-        request.context.options['connection_timeout'] = req_timeout
+            request.context.options["connection_timeout"] = min(connection_timeout, absolute_timeout)
+
+        # otherwise, try to ensure the transport's configured connection_timeout doesn't exceed absolute_timeout
+        # ("connection_config" isn't defined on HttpTransport but all implementations in this library have it)
+        elif hasattr(request.context.transport, "connection_config"):
```

Comment:
So if there is no request.context.options.connection_timeout nor transport.connection_config, timeout will not take effect?

Reply:
If we've exceeded "timeout", `absolute_timeout <= 0` and this function raised an error above. Here it's just checking whether the transport's configured connection_timeout exceeds absolute_timeout.

Reply:
The case is user-specified timeout setting is less than our default one (5 min), in this case, timeout will not take effect?

Reply:
User calls the method some_method(timeout=100)

Reply:
Assuming the transport is one of ours and configured with the default connection_timeout, when a user calls  some_method(timeout=100), this code will set connection_timeout=100 for the request.

Reply:
Yes. But if transport.connection_config does not exist (they may be not using our transport), we still want timeout setting to work, right?

Reply:
It does: RetryPolicy won't send another retry after absolute_timeout reaches 0. We make an effort to fail as soon as possible after that but don't promise to fail the moment it happens. We can't adjust connection_timeout in the case you describe, because implementing connection_timeout and any default value for it is the transport's responsibility.

Reply:
Why we cannot adjust connection_timeout?

If I am a customer, I would prefer I don't need to specify connection_timeout or read_timeout, instead, I only have timeout and SDK take care of the rest for me.

Reply:
Because implementing connection_timeout is the transport's responsibility. In your scenario, the pipeline has a custom transport that does not expose its configuration as our in-box transports do. RetryPolicy therefore can't get a connection_timeout value from the transport to compare to absolute_timeout. It can't even tell whether this transport will observe connection_timeout.

If you want the SDK to take care of these things for you, you should not use a custom transport, or your custom transport should follow our conventions 😉

Reply:
connection_timeout will be part of the contract that if users want to use their custom transport, they have to support it.

We added connection_timeout in https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/core/azure-core/README.md#configurations.

And we are working on https://github.com/Azure/azure-sdk-for-python/pull/11284.

Reply:
That doesn't force a transport to observe connection_timeout. But I don't think whether it does is crucial for the code we're commenting on. The point I'm trying to make is that if a transport implements connection_config as our transports do, then this code will set connection_timeout to fall within absolute_timeout, as you want. However, connection_config is not part of Async/HttpTransport, so a transport may not implement it. If it does not, RetryPolicy can't determine the effective connection_timeout. I think RetryPolicy should set connection_timeout only when confident that's the right thing to do.